cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/monitor/RuntimeHandler.java

```diff
@@ -85,42 +104,13 @@ public RuntimeHandler(CConfiguration cConf, MessageFetcher messageFetcher, Runna
   @POST
   @Path("/metadata")
   public void metadata(FullHttpRequest request, HttpResponder responder) throws Exception {
-    String requestBody = request.content().toString(StandardCharsets.UTF_8);
-
-    Map<String, MonitorConsumeRequest> consumeRequests = GSON.fromJson(requestBody, MAP_STRING_CONSUME_REQUEST_TYPE);
-
+    Map<Utf8, GenericRecord> consumeRequests = decodeConsumeRequest(request);
     ChunkResponder chunkResponder = responder.sendChunkStart(
-      HttpResponseStatus.OK, new DefaultHttpHeaders().set(HttpHeaderNames.CONTENT_TYPE,
-                                                          "application/json; charset=utf-8"));
+            HttpResponseStatus.OK, new DefaultHttpHeaders().set(HttpHeaderNames.CONTENT_TYPE, "avro/binary"));
     ByteBuf buffer = Unpooled.buffer();
 
-    try (JsonWriter jsonWriter = new JsonWriter(new OutputStreamWriter(new ByteBufOutputStream(buffer),
-                                                                       StandardCharsets.UTF_8))) {
-      jsonWriter.beginObject();
-
-      for (Map.Entry<String, MonitorConsumeRequest> entry : consumeRequests.entrySet()) {
-        String topicConfig = entry.getKey();
-        jsonWriter.name(topicConfig);
-        jsonWriter.beginArray();
-
-        if (!requestKeyToLocalTopic.containsKey(topicConfig)) {
-          requestKeyToLocalTopic.put(topicConfig, getTopic(topicConfig));
-        }
-
-        String topic = requestKeyToLocalTopic.get(topicConfig);
-
-        try {
-          fetchAndWriteMessages(jsonWriter, buffer, chunkResponder, topic, entry.getValue().getLimit(),
-                                entry.getValue().getMessageId());
-        } catch (Exception e) {
-          OUTAGE_LOG.error("Exception while sending messages for topic: {}", topic, e);
-        } finally {
-          jsonWriter.endArray();
-        }
-      }
-
-      jsonWriter.endObject();
-    }
+    // Encode message and send messages
+    encodeAndSendResponse(consumeRequests, chunkResponder, buffer);
```

Comment:
It's better to use a `BodyProducer` then a chunk responder. The reason is that for chunk responder, it doesn't know about the client side reading speed, hence can potentially writing too fast, causing exception raised.

Reply:
You can refer to `FetchHandler` in TMS in how to a streaming one.